framework/src/base/NonlinearSystem.C

```diff
@@ -1772,11 +1772,20 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
   jacobian.close();
 
   PARALLEL_TRY {
-    // do nodal BC
-    std::vector<numeric_index_type> zero_rows;
-
+    // We will cache NodalBC Jacobian entries here until we are ready to set them.
+    Assembly::CachedNodalBCData & cached_data = _fe_problem.assembly(0).cachedNodalBCData();
+    cached_data.clear();
+
+    // Get variable coupling list.  We do all the NodalBC stuff on
+    // thread 0...  The couplingEntries() data structure determines
+    // which variables are "coupled" as far as the preconditioner is
+    // concerned, not what variables a boundary condition specifically
+    // depends on.
+    std::vector<std::pair<MooseVariable *, MooseVariable *> > & coupling_entries = _fe_problem.couplingEntries(/*_tid=*/0);
+
+    // Compute Jacobians for NodalBCs
     ConstBndNodeRange & bnd_nodes = *_mesh.getBoundaryNodeRange();
-    for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin() ; nd != bnd_nodes.end(); ++nd)
+    for (ConstBndNodeRange::const_iterator nd = bnd_nodes.begin(); nd != bnd_nodes.end(); ++nd)
```

Comment:
In fact - thinking more about this it should be trivial to implement because each thread has an `Assembly` where it would cache the values and then during join you would join them down from one thread to the next...